core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -586,15 +583,27 @@ public RoutingNodeIterator routingNodeIter(String nodeId) {
         return nodesToShards.values().toArray(new RoutingNode[nodesToShards.size()]);
     }
 
-    public void reinitShadowPrimary(ShardRouting candidate) {
+    public ShardRouting reinitShadowPrimary(ShardRouting candidate) {
         ensureMutable();
         if (candidate.relocating()) {
             cancelRelocation(candidate);
         }
-        candidate.reinitializeShard();
+        ShardRouting reinitializedShard = candidate.reinitializeShard();
+        updateAssigned(candidate, reinitializedShard);
         inactivePrimaryCount++;
         inactiveShardCount++;
+        return reinitializedShard;
+    }
 
+    private void updateAssigned(ShardRouting oldShard, ShardRouting newShard) {
+        assert oldShard.shardId().equals(newShard.shardId());
```

Comment:
can we have some messages for these asserts, they are a pain if they trip without a message